portality/lib/dataobj.py

```diff
@@ -172,184 +178,731 @@ class DataSchemaException(Exception):
-
-        # attrs used in creating a DataObj should behave as they usually
-        # would in Python
-        if name in object.__getattribute__(self, '_RESERVED_ATTR_NAMES'):
-            return object.__getattribute__(self, name)
-
-        if object.__getattribute__(self, '_data').get(name) is None:
-        # the requested attribute is not present in the data already
-
-            if name in object.__getattribute__(self, '_struct').get('objects', []):
-                object.__getattribute__(self, '_data')[name] = DataObj(raw={}, _struct=object.__getattribute__(self, '_struct')['structs'][name], _type=name)
-            elif name in object.__getattribute__(self, '_struct').get('lists', {}):
-                object.__getattribute__(self, '_data')[name] = []
-            elif name in object.__getattribute__(self, '_struct').get('fields', {}):
-            # it's a primitive field that isn't set
-                try:
-                    # if there is a default set in the struct, return it
-                    if object.__getattribute__(self, '_struct')['fields'][name].get("default") is not None:
-                        return object.__getattribute__(self, '_struct')['fields'][name]["default"]
-                    # else cast a None
-                    return object.__getattribute__(self, '_coerce')[object.__getattribute__(self, '_struct')['fields'][name]['coerce']](None)
-                except ValueError:
-                    raise AttributeError('{name} is not set'.format(name=name))
+    def clone(self):
+        return self.__class__(deepcopy(self.data))
+
+    def json(self):
+        return json.dumps(self.data)
+
+    def _get_internal_property(self, path, wrapper=None):
+        # pull the object from the structure, to find out what kind of retrieve it needs
+        # (if there is a struct)
+        type, substruct, instructions = None, None, None
+        if self._struct:
+            type, substruct, instructions = construct_lookup(path, self._struct)
+
+        if type is None:
+            # if there is no struct, or no object mapping was found, try to pull the path
+            # as a single node (may be a field, list or dict, we'll find out in a mo)
+            val = self._get_single(path)
+
+            # if this is a dict or a list and a wrapper is supplied, wrap it
+            if wrapper is not None:
+                if isinstance(val, dict):
+                    return wrapper(val, expose_data=self._expose_data)
+                elif isinstance(val, list):
+                    return [wrapper(v, expose_data=self._expose_data) for v in val]
+
+            # otherwise, return the raw value if it is not None, or raise an AttributeError
+            if val is None:
+                raise AttributeError('{name} is not set'.format(name=path))
+
+            return val
+
+        # if the struct contains a reference to the path, always return something, even if it is None - don't raise an AttributeError
```

Comment:
and this can be set in the struct as this get__default definition I've seen, is that right?